### PR TITLE
feat: add hyperparam mutations+crossover and slightly refactor evolution

### PR DESCRIPTION
**Pull request by PatrickNercessian**

### CODE REVIEW

**Review by wmatray — COMMENTED** [review]
Looks good to me.

File: src/evaluate.py [hunk 1]
```diff
@@ -1,37 +1,42 @@
...
         data_loader: DataLoader providing training examples
         num_train_steps: Number of training steps to perform
         device: Device to train on
-        
+
     Returns:
         float: Fitness score (higher is better)
     """
-    trainer = Trainer(individual.train_config, individual.graph_module, train_data_loader)
+    copied_individual = copy.deepcopy(individual)
```
[the remark this answers is not in the document]
Comment (PatrickNercessian) [reply src/evaluate.py:30]: I think it'd be super interesting to keep training them (aside from any added nodes) in certain experiments. But I was thinking the default should just be train-from-scratch for all. What do you think?

[the remark this answers is not in the document]
Comment (wmatray) [reply src/evaluate.py:30]: I agree, just something to keep in mind. After we have working early stopping we can reconsider the ongoing training. Also I could see how ongoing training would always bias against mutations.

**Review by wmatray — APPROVED** [review]

File: src/evaluate.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (MortimerE) [reply src/evaluate.py:30]: Made a similar comment on #1 - I second keeping it like this in the default
Now that I'm thinking about it, the trained versions would have accumulated gradients in addition to trained weights, right? It might be good to clear those any time we would retrain for the undertraining or training-every-x-generations techniques we discussed 
So, even if we did keep the weights across generations for an ablation, we would probably want to ensure we're not also forwarding accumulated gradients to prevent explosion/vanishing

File: mingpt/trainer.py [hunk 2]
```diff
@@ -37,7 +37,13 @@ def __init__(self, config, model, train_dataset):
 
         # determine the device we'll train on
         if config.device == 'auto':
-            self.device = 'cuda' if torch.cuda.is_available() else 'cpu'
+            if torch.cuda.is_available():
+                self.device = 'cuda'
+            elif torch.backends.mps.is_available():
+                print("MPS available")
```
Comment (MortimerE) [inline mingpt/trainer.py:43]: just say you have an m4 max bro smh

File: src/evolution.py [hunk 3]
```diff
@@ -41,46 +52,38 @@ def run_evolution(self, num_generations: int):
...
+            for parent in self.population:  # TODO remove this
+                print(f"Parent {parent.id} has train config {parent.train_config}")
 
-        train_dataset = TextDataset(data, self.block_size)
-        # TODO: add validation dataset
-        return [self.fitness_fn(individual, train_dataset, train_dataset) for individual in self.population]
+            # Create new population through crossover and mutation
+            new_children = []
+            while len(new_children) < self.num_children_per_generation:
+                parent1, parent2 = random.sample(self.population, 2)  # TODO should this sample with or without replacement?
```
Comment (MortimerE) [inline src/evolution.py:67]: I think without replacement makes more sense, what do we gain from having children with 2 cloned parents? Or am I misreading what this does

File: src/evolution.py [hunk 4]
```diff
@@ -9,27 +9,38 @@ def __init__(
...
+            mutation_fns_and_probabilities: List of mutation functions and their respective probabilities if mutation occurs
+            crossover_fns_and_probabilities: List of crossover functions and their respective probabilities if crossover occurs
+            target_population_size: Population size to maintain after selection
+            num_children_per_generation: Number of children to generate per generation
+            block_size: Block size for the model
         """
         self.population = population
         self.fitness_fn = fitness_fn
-        self.mutation_rate = mutation_rate
-        self.selection_pressure = selection_pressure
```
Comment (MortimerE) [inline src/evolution.py:28]: why remove this? It looks like we just do top k now in _selection, don't we want to select some of the suboptimal population to be crossed and mutated?

File: src/evolution.py [hunk 5]
```diff
@@ -41,46 +52,38 @@ def run_evolution(self, num_generations: int):
...
+                print(f"Parent {parent.id} has train config {parent.train_config}")
 
-        train_dataset = TextDataset(data, self.block_size)
-        # TODO: add validation dataset
-        return [self.fitness_fn(individual, train_dataset, train_dataset) for individual in self.population]
+            # Create new population through crossover and mutation
+            new_children = []
+            while len(new_children) < self.num_children_per_generation:
+                parent1, parent2 = random.sample(self.population, 2)  # TODO should this sample with or without replacement?
+                if random.random() < self.crossover_instead_of_mutation_rate:
```
Comment (MortimerE) [inline src/evolution.py:68]: so we keep 0% of the old population?

File: src/evaluate.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (PatrickNercessian) [reply src/evaluate.py:30]: Good point

File: mingpt/trainer.py (hunk 2 above)
Comment (PatrickNercessian) [reply mingpt/trainer.py:43]: M1 pro loser

File: src/dataset.py [hunk 6]
```diff
@@ -0,0 +1,15 @@
+import torch
+
+class TextDataset(torch.utils.data.Dataset):
+    def __init__(self, data, block_size):
```
[the remark this answers is not in the document]
Comment (PatrickNercessian) [reply src/dataset.py:4]: Do what under the hood?

This is just copied from mingpt btw

File: src/evaluate.py [hunk 7]
```diff
@@ -1,11 +1,15 @@
...
+
+TOTAL_BATCHES_FOR_EVALUATION = 20
 
 def calculate_fitness(
     individual: Individual,
-    train_data_loader: torch.utils.data.DataLoader,
+    train_dataset: torch.utils.data.Dataset,
     val_data_loader: torch.utils.data.DataLoader,
     num_train_steps: int = 100,
     device: str = 'cuda' if torch.cuda.is_available() else 'cpu'
```
[the remark this answers is not in the document]
Comment (PatrickNercessian) [reply src/evaluate.py:15]: It's already passed in via the actual call to calculate_fitness

File: src/evolution.py (hunk 3 above)
Comment (PatrickNercessian) [reply src/evolution.py:67]: As in, let's say Parent 3 and Parent 146 are selected for crossover. They make a child. Then, do we want to allow Parent 146 to be selected again later for a mutation/crossover to make a new, different child?

If we do "(μ,λ)" (meaning we kill off the entire parent population), then we would indeed be halving the generation, so we wouldn't want to do it.

But I think we should do (μ+λ) (allowing selection from the parent generation), because that's typically what you do when each evaluation is pretty expensive which it is in our case (at least in the default evolution config where we do a training run for each individual).

File: src/evolution.py (hunk 4 above)
Comment (PatrickNercessian) [reply src/evolution.py:28]: It's replaced with:

```
        crossover_instead_of_mutation_rate: float = 0.5,
        mutation_fns_and_probabilities: list[tuple[Callable[[Individual], Individual], float]] = [],
        crossover_fns_and_probabilities: list[tuple[Callable[[Individual, Individual], Individual], float]] = [],
```

Comment (PatrickNercessian) [reply src/evolution.py:28]: Oh do you mean doing both a crossover and a mutation to one child?

File: src/evolution.py (hunk 5 above)
Comment (PatrickNercessian) [reply src/evolution.py:68]: No we keep both old and new, and then do selection on the concatenation of them. Unless I am missing a bug that you see

Comment (wmatray) [reply src/evolution.py:68]: If we keep every parent in self.population then line 55 will re-evaluate every parent every generation. Maybe move fitness measurement to be done in the children loop.

File: src/dataset.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (PatrickNercessian) [reply src/dataset.py:4]: Resolving - mort thought block_size referred to GPU parallelism stuff, but mingpt uses it to refer to sequence length for some reason

File: src/evolution.py (hunk 4 above)
Comment (PatrickNercessian) [reply src/evolution.py:28]: We clarified in call that we could do multiple variation operators in the same generation in ablations later.

We also clarified that selection_pressure was not removed, just replaced with 

```
        target_population_size: bool = 100,
        num_children_per_generation: int = 100,
```

**Review by MortimerE — APPROVED** [review]
nice work, impressive for someone with a broken macbook

File: src/evolution.py (hunk 5 above)
Comment (PatrickNercessian) [reply src/evolution.py:67]: https://github.com/PatrickNercessian/evolving-dnn/pull/3/commits/9fa794afa751e88e70b0af1f0474aa7f5cc6b64a

Made these changes to address the double fitness calculation. Let me know if it makes sense @MortimerE @wmatray

Comment (wmatray) [reply src/evolution.py:67]: Looks good to me
